File: backend/models/bid.py
from datetime import datetime
from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, String
from sqlalchemy.orm import relationship
from database import Base

class Bid(Base):
    __tablename__ = "bids"

    id = Column(Integer, primary_key=True, index=True)
    auction_id = Column(Integer, ForeignKey("auctions.id"), nullable=False)
    transporter_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    stage = Column(Integer, nullable=False)
    amount = Column(Float, nullable=False)
    submitted_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    status = Column(String(20), default="valid", nullable=False)

    auction = relationship("Auction", back_populates="bids")
    transporter = relationship("User")

class AuctionResult(Base):
    __tablename__ = "auction_results"

    id = Column(Integer, primary_key=True, index=True)
    auction_id = Column(Integer, ForeignKey("auctions.id"), nullable=False)
    tender_id = Column(Integer, ForeignKey("tenders.id"), nullable=False)
    winner_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    winning_bid = Column(Float, nullable=False)
    stage = Column(Integer, nullable=False, default=2)
    finalized_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    auction = relationship("Auction")
    tender = relationship("Tender")
    winner = relationship("User")

    @property
    def winner_transporter_id(self):
        return self.winner_id

    @winner_transporter_id.setter
    def winner_transporter_id(self, val):
        self.winner_id = val

    @property
    def winning_amount(self):
        return self.winning_bid

    @winning_amount.setter
    def winning_amount(self, val):
        self.winning_bid = val

    @property
    def winning_bid_id(self):
        return getattr(self, '_winning_bid_id', None)

    @winning_bid_id.setter
    def winning_bid_id(self, val):
        self._winning_bid_id = val

    @property
    def decided_at(self):
        return self.finalized_at

    @decided_at.setter
    def decided_at(self, val):
        self.finalized_at = val

File: logiq/lib/models/bid.dart
class Bid {
  final int? id;
  final int auctionId;
  final int tenderId;
  final int transporterId;
  final double amount;
  final int stage;
  final DateTime submittedAt;
  final bool isValid;

  const Bid({
    this.id,
    required this.auctionId,
    required this.tenderId,
    required this.transporterId,
    required this.amount,
    required this.stage,
    required this.submittedAt,
    this.isValid = true,
  });

  Map<String, Object?> toMap() => {
        'id': id,
        'auction_id': auctionId,
        'tender_id': tenderId,
        'transporter_id': transporterId,
        'amount': amount,
        'stage': stage,
        'submitted_at': submittedAt.toIso8601String(),
        'is_valid': isValid ? 1 : 0,
      };

  factory Bid.fromMap(Map<String, Object?> m) => Bid(
        id: m['id'] as int?,
        auctionId: (m['auction_id'] ?? 0) as int,
        tenderId: (m['tender_id'] ?? 0) as int,
        transporterId: (m['transporter_id'] ?? 0) as int,
        amount: (m['amount'] as num? ?? 0).toDouble(),
        stage: (m['stage'] ?? 1) as int,
        submittedAt:
            DateTime.tryParse((m['submitted_at'] ?? '') as String) ?? DateTime.now(),
        isValid: (m['is_valid'] as int? ?? 1) == 1,
      );
}
